python/tvm/auto_scheduler/space.py

```diff
@@ -0,0 +1,134 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+""" The class of Space used to optimize the Ansor parameters """
+
+import os
+from copy import deepcopy
+import tvm
+from tvm.auto_scheduler.measure import local_builder_build, local_run
+
+
+class MeasureResultSpace:
+    """Store the results of a measurement.
+
+    Parameters
+    ----------
+    measureResult: List[MeasureResult]
+        A List of MeasureResult.
+    """
+
+    def __init__(self, measure_result):
+        self._costs = measure_result[0].costs
+
+    @property
+    def costs(self):
+        return [v.value for v in self._costs]
+
+
+class Space:
+    """Space class
+
+    Parameters
+    ----------
+    cfg: json data
+        A json file template
+    task: SearchTask
+        The SearchTask of this measurement.
+    """
+
+    def __init__(self, cfg, task):
+        self.jfile, self.cfg = cfg, cfg["i"][1][1]
+        self.total_dims, self.dims, self.task = 0, [], task
+        self.config_space = {}
+        self.create_space()
+
+    def create_space(self):
+        """Create the space using Ansor's space"""
+        sp_space = [4, 8, 16, 24, 32, 48, 64]
+        pr_space = [64, 128, 256, 512]
+        for i in range(len(self.cfg)):
+            f = self.cfg[i]
+            if f[0] == "SP" and f[3] != 1:
+                for j in range(len(f[4])):
+                    self.config_space[f"{f[0]}_{i}_{j}"] = self.add_space(sp_space, [f[4][j]], f[3])
+            elif f[0] == "PR":
+                start_value = int(f[3].split("$")[-1])
+                if start_value != 0:
+                    self.config_space[f"{f[0]}_{i}"] = [
+                        f"auto_unroll_max_step${v}" for v in self.add_space(pr_space, [start_value])
+                    ]
+        self.dims = []
+        for key in self.config_space:
+            self.dims.append(len(self.config_space[key]))
+        self.total_dims = 1
+        if len(self.dims) > 0:
+            for dim in self.dims:
+                self.total_dims *= dim
+
+    def apply_opt(self, vals):
+        """Apply the space using Ansor's space"""
+        jfile = deepcopy(self.jfile)
```

Comment:
in `__init__` you defined self.jfile and self.cfg. I'm not sure about your intention, but maybe you can make deep copy of jfile in `__init__` and them work on.